// @openzeppelin/hardhat-upgrades/dist/admin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getManifestAdmin = exports.makeGetInstanceFunction = exports.makeTransferProxyAdminOwnership = exports.makeChangeProxyAdmin = void 0;
const upgrades_core_1 = require("@openzeppelin/upgrades-core");
const proxy_factory_1 = require("./proxy-factory");
function makeChangeProxyAdmin(hre) {
    return async function changeProxyAdmin(proxyAddress, newAdmin) {
        const admin = await getManifestAdmin(hre);
        const proxyAdminAddress = await upgrades_core_1.getAdminAddress(hre.network.provider, proxyAddress);
        if (admin.address !== proxyAdminAddress) {
            throw new Error('Proxy admin is not the one registered in the network manifest');
        }
        else if (admin.address !== newAdmin) {
            await admin.changeProxyAdmin(proxyAddress, newAdmin);
        }
    };
}
exports.makeChangeProxyAdmin = makeChangeProxyAdmin;
function makeTransferProxyAdminOwnership(hre) {
    return async function transferProxyAdminOwnership(newOwner) {
        const admin = await getManifestAdmin(hre);
        await admin.transferOwnership(newOwner);
    };
}
exports.makeTransferProxyAdminOwnership = makeTransferProxyAdminOwnership;
function makeGetInstanceFunction(hre) {
    return async function getInstance() {
        return await getManifestAdmin(hre);
    };
}
exports.makeGetInstanceFunction = makeGetInstanceFunction;
async function getManifestAdmin(hre) {
    const manifest = await upgrades_core_1.Manifest.forNetwork(hre.network.provider);
    const manifestAdmin = await manifest.getAdmin();
    const proxyAdminAddress = manifestAdmin === null || manifestAdmin === void 0 ? void 0 : manifestAdmin.address;
    if (proxyAdminAddress === undefined) {
        throw new Error('No ProxyAdmin was found in the network manifest');
    }
    const AdminFactory = await proxy_factory_1.getProxyAdminFactory(hre);
    return AdminFactory.attach(proxyAdminAddress);
}
exports.getManifestAdmin = getManifestAdmin;
